pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexReaderFactory.java

```diff
@@ -45,6 +45,18 @@ R createIndexReader(SegmentDirectory.Reader segmentReader, FieldIndexConfigs fie
     protected abstract R createIndexReader(PinotDataBuffer dataBuffer, ColumnMetadata metadata, C indexConfig)
         throws IOException, IndexReaderConstraintException;
 
+    /**
+     * Sometimes the index configuration indicates that the index should be disabled but the reader actually contains
+     * a buffer for the index type.
+     *
+     * By default, the buffer has priority over the configuration, so in case we have a buffer we would create an index
```

Comment:
Don't think this is the case. Originally, (https://github.com/apache/pinot/pull/10184/files#diff-271b081664e0c34f39b5ee282f33dc95bcc6cc45e94daa0aa89e8b6bc53213d6L82-L89)
most index types would be config > buffer except a few (like dictionary).

Reply:
But do we want to keep the previous behavior or do we want to be resilient in cases where the table config says one thing and the metainfo says otherwise? If we are going to decide what to do in the general case (in IndexType) based on what PhysicalColumnIndexContainer did, we should offer the two options (use config or use metainfo) and let the caller decide.

Reply:
In typical pinot server loading a segment case, I honestly don't see why it would be a problem. When servers come up, If there's a mismatch b/w table config and local segment metadata, the segment is always preprocessed as per the latest config before loading. So post preprocessing, the segment metadata should always be the source of truth. i.e. you're right that at the IndexReader level, we can always give preference to buffer over table config.

Although I'd let @Jackie-Jiang  / @npawar / @klsince share their thoughts.

Reply:
What about this approach? Now the decision between prioritizing buffers or not is delegated to the caller. By default indexes will honor the config, but dictionary and null value will not.

I've also found that PhysicalColumnIndexContainer does not load range index when the column is sorted, while RangeIndexType does. I'm going to leave it as it is right now. I don't think we should add that condition into RangeIndexType and I would prefer to do not add a special case in PhysicalColumnIndexContainer. I assume the reason why PhysicalColumnIndexContainer did that was because it wasn't actually necessary to use the range index in that situation, but I guess that should be a responsibility of the planer.

Reply:
Sorted column config is actually quite tricky. It is used when creating the segment (on the minion side, or when sealing the consuming segment), but once the segment is created, it is no longer honored, and we rely on the metadata to determine whether a column is sorted. There could be multiple columns sorted, but they might not be configured as sorted column.
When a column is sorted, but configured as no-dictionary column, we choose to ignore the no-dictionary config because it is almost always more efficient to use dictionary encoding for sorted column. The most common scenario would be user added a new column to an existing table, and wants it to be no-dictionary. For the existing segments, Pinot will backfill default value for the new column, so it contains only one value thus being sorted. In such case, we don't want to use no-dictionary because that won't be efficient.